src/main/java/co/rsk/federate/watcher/FederationWatcher.java

```diff
@@ -0,0 +1,124 @@
+        this.rsk = rsk;
+    }
+
+    /**
+     * Starts the {@code FederationWatcher} by setting the {@code FederationProvider}
+     * and {@code FederationWatcherListener}, and begins listening for new blocks on
+     * the RSK blockchain.
+     *
+     * @param federationProvider        the provider used to obtain federation information (active and retiring federations)
+     * @param federationWatcherListener the listener that will be notified when the federations change
+     */
+    public void start(FederationProvider federationProvider, FederationWatcherListener federationWatcherListener) {
+        this.federationProvider = federationProvider;
+        this.federationWatcherListener = federationWatcherListener;
+     
+        rsk.addListener(new EthereumListenerAdapter() {
+            @Override
+            public void onBestBlock(org.ethereum.core.Block block, List<TransactionReceipt> receipts) {
+                // Updating state only when the best block changes still "works",
+                // since we're interested in finding out only when the active or retiring federation(s) changed.
+                // If there was a side chain in which any of these changed in, say, block 4500, but
+                // that side chain became main chain in block 7800, it would still be ok to
+                // start monitoring the new federation(s) on that block.
+                // The main reasons for this is that RSK nodes would have never reported the active or
+                // retiring federation(s) being different *before* the best chain change. Therefore,
+                // there should be no new Bitcoin transactions directed to these new addresses
+                // until this change effectively becomes a part of RSK's "reality".
+                // The case in which we go back and forth to and from a sidechain in which the
+                // federation effectively changed is still to be explored deeply, but the same reasoning
+                // should apply since going back and forth would trigger two federation changes.
+                // A client trying to send bitcoins to the new federation without waiting
+                // a good number of confirmations would be, essentially, "playing with fire".
+                logger.info("[updateState] New best block, updating state");
+                updateState();
+            }
+        });
+    }
+
+    /**
+     * Updates the current state of the federations by checking if the active or
+     * retiring federations have changed. If a federation change is detected, it notifies
+     * the {@code FederationWatcherListener}.
+     */
+    private void updateState() {
+        updateActiveFederation();
+        updateRetiringFederation();
+    }
+
+    private void updateActiveFederation() {
+        // Check if active federation has changed
+        Optional<Address> currentlyActiveFederationAddress = Optional.ofNullable(federationProvider.getActiveFederationAddress());
```

Comment:
why optional? shouldn't always be an active federation?

Reply:
This is to do the equality check between the optionals.

Reply:
The issue arises if for instance activeFederation is null. This line `Address oldActiveFederationAddress = activeFederation.getAddress();`,  could throw an NPE if activeFederation is null. Wrapping in an Optional avoids this case. Federation provider should always return a not null federation, but we still wrap it in an Optional just to make it consistent with the Retiring Federation implementation which is not always the case.

Reply:
i still don't like it. If is null, we should throw an exception (or just return with a logged error) either before passing the value or in this method. I would avoid ending with a null active federation (not sure if this could happen but)

Reply:
You can have an active federation as null at node start time. I have changed it to the original implementation. Let me know if this is fine. Thanks!